frontend/shared/utils/_helpers.js

```diff
@@ -5,3 +5,38 @@ export const addOuterTag = function(html, tag) {
         match = html.match(regex);
     return match == null || match[1] != tag ? `<${tag}>${html}</${tag}>` : html;
 };
+
+export const markKeywords = function(text, allSettings) {
+    const all_tokens = allSettings.set1.keywords
+            .concat(allSettings.set2.keywords)
+            .concat(allSettings.set3.keywords),
+        all_re = new RegExp(
+            "\\b".concat(all_tokens.join("\\b|\\b").replace(/\*/g, "[\\S]*?")).concat("\\b"),
+            "gim"
+        );
+    if (all_tokens.length === 0) {
+        return text;
+    }
+    text = text.replace(all_re, match => `<mark>${match}</mark>`);
+    text = markText(text, allSettings.set1);
+    text = markText(text, allSettings.set2);
+    text = markText(text, allSettings.set3);
+    return text;
+};
+
+export const markText = function(text, settings) {
+    if (settings.keywords.length === 0) {
+        return text;
+    }
+    const re = new RegExp(
+        `<mark>(?<token>\\b${settings.keywords
+            .join("\\b|\\b")
+            .replace(/\*/g, "[\\S]*?")}\\b)</mark>`,
```

Comment:
This pattern is different than the all_re pattern above - I think that above one that concats the \b at the beginning and end seems better?   Could we make a n few function that build the regex given an array of keywords, that's used everywhere?

Reply:
the first regex matches just the keywords, and it's used to add the <mark> elements. this regex matches keywords with the <mark> element (and the beginning and end \b tokens are present, just bookended by the <mark> elements). we did this so users could use 'mark', 'positive', 'negative', etc. as keywords without creating regex conflicts. So I think these regex need to be different